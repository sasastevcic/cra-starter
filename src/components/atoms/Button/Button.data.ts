import { ReactNode } from 'react';
import { NativeProps } from '../../atoms/AbstractButton/AbstractButton.data';

export enum ButtonTheme {
	Dark = 'dark',
	Light = 'light',
}

export interface StyledButtonProps {
	$buttonTheme: ButtonTheme;
	$onlyIcon: boolean;
}

export interface ButtonProps extends Omit<NativeProps, 'ref'> {
	icon?: ReactNode;
	buttonTheme?: ButtonTheme;
}
