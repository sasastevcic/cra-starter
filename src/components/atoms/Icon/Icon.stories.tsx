import { ReactElement } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Table as GenericTable } from '../../../styles/storybook/Storybook';
import { StyledRow, StyledData, StyledIcon } from '../../../styles/storybook/Storybook.styles';
import Paragraph from '../Paragraph';
import { Icon } from './Icon';

const Table = (): ReactElement => {
	const tableHeadings = ['Name', 'Preview'];

	const rows = Object.entries(Icon).map(([name, Svg]) => (
		<StyledRow key={name}>
			<StyledData>
				<Paragraph>{name}</Paragraph>
			</StyledData>
			<StyledData>
				<StyledIcon as={Svg} />
			</StyledData>
		</StyledRow>
	));

	return <GenericTable headings={tableHeadings}>{rows}</GenericTable>;
};

export default {
	title: 'Vector/Icons',
	component: Table,
} as ComponentMeta<typeof Table>;

const Template: ComponentStory<typeof Table> = () => <Table />;

export const Icons = Template.bind({});
