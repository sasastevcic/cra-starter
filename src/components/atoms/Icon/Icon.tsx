import { ReactComponent as Close } from './svg/close.svg';
import { ReactComponent as Logo } from './svg/logo.svg';

export const Icon = {
	Close,
	Logo,
};
