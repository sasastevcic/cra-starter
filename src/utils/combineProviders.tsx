/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
