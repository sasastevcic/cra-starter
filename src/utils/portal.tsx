import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
