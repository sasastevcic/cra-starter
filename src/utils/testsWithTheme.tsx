/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
