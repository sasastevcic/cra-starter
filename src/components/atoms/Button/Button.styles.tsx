import { size } from 'polished';
import styled, { css, DefaultTheme, FlattenInterpolation, ThemeProps } from 'styled-components';
import { typeStyle } from '../../../styles/config/typeStyles';
import { FontWeight, TransitionDuration, TransitionEase } from '../../../styles/config/variables';
import { hover } from '../../../styles/helpers/hover';
import { AbstractButton } from '../../atoms/AbstractButton/AbstractButton';
import { ButtonTheme, StyledButtonProps } from './Button.data';

const SIZE = 4;

const darkHover = css`
	background-color: ${({ theme }) => theme.color.blackFaded};
`;

const lightHover = css`
	background-color: ${({ theme }) => theme.color.whiteFaded};
`;

const dark = css`
	background-color: ${({ theme }) => theme.color.black};
	color: ${({ theme }) => theme.color.white};

	${hover(darkHover)}
`;

const light = css`
	background-color: ${({ theme }) => theme.color.white};
	color: ${({ theme }) => theme.color.black};

	${hover(lightHover)}
`;

const disabled = css`
	background-color: ${({ theme }) => theme.color.blackFaded};
	color: ${({ theme }) => theme.color.whiteFaded};
	cursor: not-allowed;
`;

const onlyIcon = css`
	${size(`${SIZE}rem`)};
	min-height: auto;
	border-radius: 50%;
	padding: 0;

	svg {
		margin-left: 0;
	}
`;

const themes: { [key in ButtonTheme]: FlattenInterpolation<ThemeProps<DefaultTheme>> } = {
	dark,
	light,
};

export const StyledButton = styled(AbstractButton)<StyledButtonProps>`
	${typeStyle.normal};
	font-weight: ${FontWeight.Bold};
	display: inline-flex;
	align-items: center;
	justify-content: center;
	padding: 0.2rem 2.5rem;
	min-height: ${SIZE}rem;
	border-radius: 4.6rem;
	transition: ${TransitionDuration.Normal} ${TransitionEase.Default};
	transition-property: color, background-color;

	${({ $buttonTheme }) => themes[$buttonTheme]};

	&:disabled {
		${disabled};
	}

	svg {
		${size(`${SIZE * 0.4}rem`)};
		margin-left: 1rem;
	}

	${({ $onlyIcon }) => $onlyIcon && onlyIcon}
`;
