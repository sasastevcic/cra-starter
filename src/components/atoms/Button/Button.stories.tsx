import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Icon } from '../../atoms/Icon/Icon';
import { ButtonTheme } from './Button.data';
import Button from './index';

export default {
	title: 'Form Elements/Buttons',
	component: Button,
	args: {
		children: 'Button',
	},
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

const darkTheme = {
	buttonTheme: ButtonTheme.Dark,
};

const lightTheme = {
	buttonTheme: ButtonTheme.Light,
};

export const Dark = Template.bind({});
Dark.args = darkTheme;

export const DarkWithIcon = Template.bind({});
DarkWithIcon.args = {
	...darkTheme,
	icon: <Icon.Logo />,
};

export const DarkIcon = Template.bind({});
DarkIcon.args = {
	...darkTheme,
	icon: <Icon.Logo />,
	children: null,
};

export const Light = Template.bind({});
Light.args = lightTheme;
Light.parameters = {
	backgrounds: { default: 'dark' },
};

export const LightWithIcon = Template.bind({});
LightWithIcon.args = {
	...lightTheme,
	icon: <Icon.Logo />,
};
LightWithIcon.parameters = {
	backgrounds: { default: 'dark' },
};

export const LightIcon = Template.bind({});
LightIcon.args = {
	...lightTheme,
	icon: <Icon.Logo />,
	children: null,
};
LightIcon.parameters = {
	backgrounds: { default: 'dark' },
};
