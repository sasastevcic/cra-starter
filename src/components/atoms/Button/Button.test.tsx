import renderer from 'react-test-renderer';
import { renderWithTheme, componentWithTheme } from '../../../utils/testsWithTheme';
import { Icon } from '../../atoms/Icon/Icon';
import Button from './index';

describe('Button', () => {
	describe('Render', () => {
		it('should render', () => {
			const { getByTestId } = renderWithTheme(<Button />);

			const element = getByTestId('Button');

			expect(element).toBeInTheDocument();
		});

		it('should render an icon', () => {
			const { getByTestId } = renderWithTheme(<Button icon={<Icon.Logo />} />);

			const element = getByTestId('Button');
			const child = element.querySelector('svg');

			expect(element).toContainElement(child);
		});

		it('should be an anchor tag if there is a href prop', () => {
			const { getByTestId } = renderWithTheme(<Button href="https://www.google.com" />);

			const element = getByTestId('Button');

			expect(element).toContainHTML('</a>');
			expect(element).toHaveAttribute('href', 'https://www.google.com');
		});

		it('should be button tag by default', () => {
			const { getByTestId } = renderWithTheme(<Button />);

			const element = getByTestId('Button');

			expect(element).toContainHTML('</button>');
		});

		it('should render correctly', () => {
			const tree = renderer.create(componentWithTheme(<Button />)).toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Content', () => {
		it('should have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Button>
					<span>Child</span>
				</Button>,
			);

			const element = getByTestId('Button');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});

	describe('Attributes', () => {
		it('should be disabled', () => {
			const { getByTestId } = renderWithTheme(<Button disabled />);

			const element = getByTestId('Button');

			expect(element).toBeDisabled();
		});
	});
});
