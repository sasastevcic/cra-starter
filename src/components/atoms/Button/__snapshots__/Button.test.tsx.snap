// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render should render correctly 1`] = `
<button
  className="sc-bczRLJ hPaJzN"
  data-testid="Button"
  onClick={[Function]}
  type="button"
/>
`;
